test/e2e/storage/static_pods.go

```diff
@@ -0,0 +1,184 @@
+			}
+		})
+
+		ginkgo.By("Stop the control plane by removing the static pod manifests")
+		err = stopControlPlane(ctx)
+		framework.ExpectNoError(err, "Stopping the control plane")
+
+		ginkgo.By("Wait for the API server to disappear")
+		gomega.Eventually(ctx, func(ctx context.Context) error {
+			return checkAPIServer(ctx, false, f.ClientSet)
+		}).WithTimeout(time.Minute).Should(gomega.Succeed())
+
+		ginkgo.By("Stop kubelet")
+		err = stopKindKubelet(ctx)
+		framework.ExpectNoError(err, "Stopping kubelet")
+
+		ginkgo.By("Re-enable control plane")
+		err = enableControlPlane(ctx)
+		framework.ExpectNoError(err, "Re-enabling control plane by restoring the static pod manifests")
+
+		ginkgo.By("Start kubelet")
+		err = startKindKubelet(ctx)
+		framework.ExpectNoError(err, "Starting kubelet")
+
+		ginkgo.By("Wait for a static pod with the API server to start")
+		const apiServerTimeout = 10 * time.Minute // It takes up to 2 minutes on my machine
+		gomega.Eventually(ctx, func(ctx context.Context) error {
+			return checkAPIServer(ctx, true, f.ClientSet)
+		}).WithTimeout(apiServerTimeout).Should(gomega.Succeed())
+	})
+})
+
+func stopKindKubelet(ctx context.Context) error {
+	return kubeletExec("systemctl", "stop", "kubelet")
+}
+
+func startKindKubelet(ctx context.Context) error {
+	return kubeletExec("systemctl", "start", "kubelet")
+}
+
+func stopControlPlane(ctx context.Context) error {
+	// Move all static pod manifests to a backup dir (incl. KCM, scheduler, API server and etcd)
+	return kubeletExec("sh", "-c", "mkdir -p /etc/kubernetes/manifests-backup && mv /etc/kubernetes/manifests/* /etc/kubernetes/manifests-backup/")
+}
+
+func enableControlPlane(ctx context.Context) error {
+	// Restore the static pod manifests from the backup dir.
+	// Using `sh` to get `*` expansion.
+	return kubeletExec("sh", "-c", "mv /etc/kubernetes/manifests-backup/* /etc/kubernetes/manifests/")
+}
+
+// Run a command in container with kubelet (and the whole control plane as containers)
+func kubeletExec(command ...string) error {
+	containerName := getKindContainerName()
+	args := []string{"exec", containerName}
+	args = append(args, command...)
+	cmd := exec.Command("docker", args...)
+
+	out, err := cmd.CombinedOutput()
+	if err != nil {
+		return fmt.Errorf("command %q failed: %v\noutput:%s", prettyCmd(cmd), err, string(out))
+	}
+
+	framework.Logf("command %q succeeded:\n%s", prettyCmd(cmd), string(out))
+	return nil
+}
+
+func getKindContainerName() string {
+	clusterName := os.Getenv("KIND_CLUSTER_NAME")
+	if clusterName == "" {
+		clusterName = "kind"
+	}
+	return clusterName + "-control-plane"
+}
+
+func prettyCmd(cmd *exec.Cmd) string {
+	return fmt.Sprintf("%s %s", cmd.Path, strings.Join(cmd.Args, " "))
+}
+
+func checkAPIServer(ctx context.Context, expectedAvailable bool, cs clientset.Interface) error {
+	// Use a short timeout to check more often.
+	ctx, done := context.WithTimeout(ctx, 5*time.Second)
+	defer done()
```

Comment:
You could use `Eventually().WithPolling(5 * time.Second)` for this?

Reply:
I wanted `List()` below to timeout quickly if the API server is not available, `ctx` would never stop (unless terminated). I can see there is a `metav1.ListOptions{Timeout: *int64}` for that, so I'm using it now.